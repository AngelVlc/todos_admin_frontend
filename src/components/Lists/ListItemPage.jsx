import React, { useEffect, useState } from 'react';
import { useParams, useHistory } from 'react-router-dom';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import axios from 'axios';
import * as Yup from 'yup';

export const ListItemPage = () => {
    let { listId, itemId } = useParams();
    let history = useHistory();

    const initialState = {
        pageTitle: 'New item',
        submintBtnText: 'CREATE',
        submitUrl: `lists/${listId}/items`,
        isNew: true,
        title: '',
        description: ''
    }
    const [pageState, setPageState] = useState(initialState);

    useEffect(() => {
        const getExistingItem = async () => {
            const res = await axios.get(`lists/${listId}/items/${itemId}`)
            const item = res.data

            setPageState({
                pageTitle: `Edit item '${item.title}'`,
                submintBtnText: 'SAVE',
                submitUrl: `lists/${listId}/items/${itemId}`,
                isNew: false,
                title: item.title,
                description: item.description
            });
        }
        if (listId && itemId) {
            getExistingItem();
        }
    }, [listId, itemId]);

    return (
        <div className="container">
            <h3 className="title">{pageState.pageTitle}</h3>
            <Formik
                enableReinitialize={true}
                initialValues={{
                    title: pageState.title,
                    description: pageState.description
                }}
                validationSchema={
                    Yup.object({
                        title: Yup.string()
                            .required('Required')
                    })
                }
                onSubmit={async ({ title, description }) => {
                    const body = {
                        title,
                        description
                    }
                    if (pageState.isNew) {
                        await axios.post(pageState.submitUrl, body)
                    } else {
                        await axios.put(pageState.submitUrl, body)
                    }
                    history.push(`/lists/${listId}/edit`);
                }}>
                <Form>
                    <div className="field">
                        <label className="label" htmlFor="title">Title</label>
                        <div className="control">
                            <Field name="title" type="text" data-testid="title" />
                        </div>
                        <p className="help is-danger" data-testid="titleErrors">
                            <ErrorMessage name="title" />
                        </p>
                    </div>

                    <div className="field">
                        <label className="label" htmlFor="description">Description</label>
                        <div className="control">
                            <Field name="description" type="text" data-testid="description" />
                        </div>
                        <p className="help is-danger" data-testid="descriptionErrors">
                            <ErrorMessage name="description" />
                        </p>
                    </div>

                    <div className="field is-grouped">
                        <p className="control">
                            <button className="button" data-testid="submit" type="submit">{pageState.submintBtnText}</button>
                        </p>
                        <p className="control">
                            <button className="button" data-testid="cancel" type="button" onClick={() => history.push(`/lists/${listId}/edit`)}>CANCEL</button>
                        </p>
                        {!pageState.isNew &&
                            <p className="control">
                                <button className="button is-danger" data-testid="delete" type="button" onClick={() => history.push(`/lists/${listId}/items/${itemId}/delete`)}>DELETE</button>
                            </p>
                        }
                    </div>
                </Form>
            </Formik>
        </div>
    )
}
