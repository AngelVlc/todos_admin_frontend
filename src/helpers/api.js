import { requestPending, requestDone, requestFailed } from '../actions'

const backendUrl = process.env.REACT_APP_BACKEND_URL || 'http://localhost:5001'

export const doGetToken = (loginDto, requestsDispatch) => {
    return new Promise((resolve, reject) => {
        requestsDispatch(requestPending());
        const requestOptions = {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(loginDto)
        };
        fetch(`${backendUrl}/auth/token`, requestOptions)
            .then(res => {
                requestsDispatch(requestDone());
                if (!res.ok) {
                    res.text().then(txt => reject(txt));
                    return
                }
                res.json().then(obj => {
                    resolve(obj);
                })
            })
            .catch(error => {
                requestsDispatch(requestDone());
                reject(error.message);
            })
    });
}

const getHeaders = (method, token) => {
    return {
        method: method,
        headers: { 'Content-Type': 'application/json', 'Authorization': `Bearer ${token}` },
    };
}

export const doGet = (endpoint, token, requestsDispatch) => {
    return new Promise((resolve, reject) => {
        requestsDispatch(requestPending());
        const requestOptions = getHeaders('GET', token);
        fetch(`${backendUrl}/${endpoint}`, requestOptions)
            .then(res => {
                requestsDispatch(requestDone());
                if (res.ok) {
                    res.json().then(obj => {
                        resolve(obj);
                    })
                } else {
                    res.text().then(txt => {
                        requestsDispatch(requestFailed(txt));
                        reject(txt)
                    });
                }
            })
            .catch(error => {
                requestsDispatch(requestFailed(error.message));
                reject(error.message);
            })
    });
}

export const doDelete = (endpoint, token, requestsDispatch) => {
    return new Promise((resolve, reject) => {
        requestsDispatch(requestPending());
        const requestOptions = getHeaders('DELETE', token);
        fetch(`${backendUrl}/${endpoint}`, requestOptions)
            .then(res => {
                requestsDispatch(requestDone());
                if (res.ok) {
                    resolve();
                    return
                }
                res.text().then(txt => {
                    requestsDispatch(requestFailed(txt));
                    reject(txt)
                });
            })
            .catch(error => {
                requestsDispatch(requestFailed(error.message));
                reject(error.message);
            })
    });
}

export const doPost = (endpoint, body, token, requestsDispatch) => {
    return new Promise((resolve, reject) => {
        requestsDispatch(requestPending());
        const requestOptions = getHeaders('POST', token);
        requestOptions.body = JSON.stringify(body);
        fetch(`${backendUrl}/${endpoint}`, requestOptions)
            .then(res => {
                requestsDispatch(requestDone());
                res.text().then(txt => {
                    if (res.ok) {
                        resolve(txt);
                        return
                    }
                    requestsDispatch(requestFailed(txt));
                    reject(txt)
                });
            })
            .catch(error => {
                requestsDispatch(requestFailed(error.message));
                reject(error.message);
            })
    });
}
