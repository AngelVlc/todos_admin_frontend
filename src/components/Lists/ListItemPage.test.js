import { render, cleanup, fireEvent, wait } from '@testing-library/react'
import { ListItemPage } from './ListItemPage'
import { AppContext } from '../../contexts/AppContext'
import * as api from '../../helpers/api';
import { MemoryRouter, Route } from 'react-router-dom'
import { act } from 'react-dom/test-utils';

jest.mock('../../helpers/api');
const mockHistoryGoBack = jest.fn();
const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
        goBack: mockHistoryGoBack,
        push: mockHistoryPush
    })
}));

const renderWithContextAndRouterForExistingItem = (component) => {
    api.doGet.mockResolvedValue({
        id: 2,
        title: 'item title',
        description: 'item description'
    });
    const context = { auth: { info: {} } };
    return {
        ...render(
            <AppContext.Provider value={context}>
                <MemoryRouter initialEntries={[`/lists/2/items/5/edit`]}>
                    <Route path="/lists/:listId/items/:itemId/edit">
                        {component}
                    </Route>
                </MemoryRouter>
            </AppContext.Provider>)
    }
}

const renderWithContextAndRouterForNewItem = (component) => {
    const context = { auth: { info: {} } };
    return {
        ...render(
            <AppContext.Provider value={context}>
                <MemoryRouter initialEntries={[`/lists/2/items/new`]}>
                    <Route path="/lists/:listId/items/new">
                        {component}
                    </Route>
                </MemoryRouter>
            </AppContext.Provider>)
    }
}

afterEach(cleanup)

it('should match the snapshot for an existing item', async () => {
    let fragment;
    await act(async () => {
        const { asFragment } = renderWithContextAndRouterForExistingItem(<ListItemPage />);
        fragment = asFragment;
    });
    expect(fragment(<ListItemPage />)).toMatchSnapshot();
});

it('should match the snapshot for a new item', async () => {
    const { asFragment } = renderWithContextAndRouterForNewItem(<ListItemPage />);
    expect(asFragment(<ListItemPage />)).toMatchSnapshot();
});

it('should allow delete an existing item', async () => {
    let container
    await act(async () => {
        container = renderWithContextAndRouterForExistingItem(<ListItemPage />);
    });

    await wait(() => {
        fireEvent.click(container.getByTestId('delete'));
    })

    expect(mockHistoryPush.mock.calls.length).toBe(1);
    expect(mockHistoryPush.mock.calls[0][0]).toBe('/lists/2/items/5/delete');
    mockHistoryPush.mockClear();
});

it('should allow cancel', async () => {
    const { getByTestId } = renderWithContextAndRouterForNewItem(<ListItemPage />);

    await wait(() => {
        fireEvent.click(getByTestId('cancel'));
    })

    expect(mockHistoryGoBack.mock.calls.length).toBe(1);
    mockHistoryGoBack.mockClear();
});

it('should require item title', async () => {
    const { getByTestId } = renderWithContextAndRouterForNewItem(<ListItemPage />);

    await wait(() => {
        fireEvent.click(getByTestId('submit'));
    })

    expect(getByTestId('titleErrors')).toHaveTextContent('Required');
})

it('should update an existing item', async () => {
    let container
    await act(async () => {
        container = renderWithContextAndRouterForExistingItem(<ListItemPage />);
    });

    await changeInputValue(container.getByTestId, 'title', 'updated title');
    await changeInputValue(container.getByTestId, 'description', 'updated description');

    api.doPut.mockResolvedValue({});

    await wait(() => {
        fireEvent.click(container.getByTestId('submit'));
    })

    expect(api.doPut.mock.calls.length).toBe(1);
    expect(api.doPut.mock.calls[0][0]).toBe('lists/2/items/5');
    expect(api.doPut.mock.calls[0][1]).toStrictEqual({ title: 'updated title', description: 'updated description' });

    expect(mockHistoryGoBack.mock.calls.length).toBe(1);
    mockHistoryGoBack.mockClear();
});

it('should create a new item', async () => {
    const { getByTestId } = renderWithContextAndRouterForNewItem(<ListItemPage />);

    await changeInputValue(getByTestId, 'title', 'title');
    await changeInputValue(getByTestId, 'description', 'description');

    api.doPost.mockResolvedValue({});

    await wait(() => {
        fireEvent.click(getByTestId('submit'));
    })

    expect(api.doPost.mock.calls.length).toBe(1);
    expect(api.doPost.mock.calls[0][0]).toBe('lists/2/items');
    expect(api.doPost.mock.calls[0][1]).toStrictEqual({ title: 'title', description: 'description' });

    expect(mockHistoryGoBack.mock.calls.length).toBe(1);
    mockHistoryGoBack.mockClear();
});

const changeInputValue = async (getByTestId, name, value) => {
    await wait(() => {
        fireEvent.change(getByTestId(name), {
            target: {
                value: value
            }
        })
    })
}

