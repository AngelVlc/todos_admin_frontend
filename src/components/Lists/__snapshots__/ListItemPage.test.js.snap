// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot for a new item 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h3
      class="title"
    >
      New item
    </h3>
    <nav
      aria-label="breadcrumbs"
      class="breadcrumb"
    >
      <ul>
        <li>
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            href="/lists"
          >
            Lists
          </a>
        </li>
        <li>
          <a
            href="/lists/2/edit"
          >
            List
          </a>
        </li>
        <li
          class="is-active"
        >
          <a
            aria-current="page"
            href="/lists/2/items/undefined/edit"
          >
            new
          </a>
        </li>
      </ul>
    </nav>
    <form
      action="#"
    >
      <div
        class="field"
      >
        <label
          class="label"
          for="title"
        >
          Title
        </label>
        <div
          class="control"
        >
          <input
            data-testid="title"
            name="title"
            type="text"
            value=""
          />
        </div>
        <p
          class="help is-danger"
          data-testid="titleErrors"
        />
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="description"
        >
          Description
        </label>
        <div
          class="control"
        >
          <input
            data-testid="description"
            name="description"
            type="text"
            value=""
          />
        </div>
        <p
          class="help is-danger"
          data-testid="descriptionErrors"
        />
      </div>
      <div
        class="field is-grouped"
      >
        <p
          class="control"
        >
          <button
            class="button"
            data-testid="submit"
            type="submit"
          >
            CREATE
          </button>
        </p>
        <p
          class="control"
        >
          <button
            class="button"
            data-testid="cancel"
            type="button"
          >
            CANCEL
          </button>
        </p>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`should match the snapshot for an existing item 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h3
      class="title"
    >
      Edit item 'item title'
    </h3>
    <nav
      aria-label="breadcrumbs"
      class="breadcrumb"
    >
      <ul>
        <li>
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            href="/lists"
          >
            Lists
          </a>
        </li>
        <li>
          <a
            href="/lists/2/edit"
          >
            List
          </a>
        </li>
        <li
          class="is-active"
        >
          <a
            aria-current="page"
            href="/lists/2/items/5/edit"
          >
            item title
          </a>
        </li>
      </ul>
    </nav>
    <form
      action="#"
    >
      <div
        class="field"
      >
        <label
          class="label"
          for="title"
        >
          Title
        </label>
        <div
          class="control"
        >
          <input
            data-testid="title"
            name="title"
            type="text"
            value="item title"
          />
        </div>
        <p
          class="help is-danger"
          data-testid="titleErrors"
        />
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="description"
        >
          Description
        </label>
        <div
          class="control"
        >
          <input
            data-testid="description"
            name="description"
            type="text"
            value="item description"
          />
        </div>
        <p
          class="help is-danger"
          data-testid="descriptionErrors"
        />
      </div>
      <div
        class="field is-grouped"
      >
        <p
          class="control"
        >
          <button
            class="button"
            data-testid="submit"
            type="submit"
          >
            SAVE
          </button>
        </p>
        <p
          class="control"
        >
          <button
            class="button"
            data-testid="cancel"
            type="button"
          >
            CANCEL
          </button>
        </p>
        <p
          class="control"
        >
          <button
            class="button is-danger"
            data-testid="delete"
            type="button"
          >
            DELETE
          </button>
        </p>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
